Fix dataarray drop attrs

j-haacker opened the pull request:
This PR is related to Issue #10027

It fixes the issue and adds tests to prevent similar issues in future.

- [x] Closes #10027
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

I did not document the change in whats-new.

I'm also not 100 % happy with the solution. I would welcome an experienced contributor to review the changes.

headtr1ck commented on xarray/core/dataarray.py:
```diff
@@ -7575,6 +7575,13 @@ def drop_attrs(self, *, deep: bool = True) -> Self:
-        )
+        if not deep:
+            for k in list(self.attrs):
+                del self.attrs[k]
```
That changes the data array inplace which is something we try to avoid.
Can you return a copy instead while leaving the original one intact?

j-haacker replied:
Thank you, I wasn't aware. I did so with `self_copy = self.copy()`.

Further, I tidied up the commit history, such that now it is only the "Fix ..." and the "Add tests ..." with the pre-commit changes already included.

headtr1ck replied:
Maybe there is a nicer way with `_replace`, but not sure.

Don't worry about the commits, we squash merge to main anyway.

j-haacker replied:
@mathause tried `_replace()` but noted that it doesn't work on attributes (see https://github.com/pydata/xarray/issues/10027#issuecomment-2638937097).

I do think, though, there might be a more elegant solution. Splitting up the two cases (`deep`) and treating them entirely different, to me looks like predefined future issues. However, I couldn't think of a better fix.

max-sixty commented:
A welcome follow-up would be adding `attrs` to `_replace`, like we have in Dataset. I'm not sure why it's missing tbh. But no obligation.

Could we add a whatsnew and then merge?

j-haacker commented:
I added attrs to `_replace` by replacing the attrs of the variable because attrs cannot be passed to the `DataArray` constructor by themselves.

I also updated the `whats-new.rst`.

max-sixty reviewed (approved):
Excellent! Thanks!

max-sixty commented:
@headtr1ck  you need to approve for us to merge

headtr1ck commented on xarray/core/dataarray.py:
```diff
@@ -532,6 +534,11 @@ def _replace(
             indexes = self._indexes
         if name is _default:
             name = self.name
+        if attrs is _default:
```
Could have gone with None, because None is not  a valid attrs type.
But ok, doesn't hurt.

j-haacker replied:
I'm not entirely sure here: would it be favorable to use `None`? I don't see a problem and will be happy to change that.

headtr1ck commented on xarray/core/dataarray.py:
```diff
@@ -532,6 +534,11 @@ def _replace(
+        if attrs is _default:
+            attrs = copy.copy(self.attrs)
+        else:
+            variable = self.variable.copy()
```
I don't think this is correct.
This will overwrite a given variable.

j-haacker replied:
Yes, true, I'll fix this.

max-sixty replied:
thanks a lot @headtr1ck , sorry to have missed this

headtr1ck commented on xarray/core/dataarray.py:
```diff
@@ -532,6 +534,11 @@ def _replace(
+            attrs = copy.copy(self.attrs)
+        else:
+            variable = self.variable.copy()
+            variable.attrs = attrs
```
Is there a reason you add the attributes to the variable instead of passing them to the constructor below?
Seems much easier to do so.

j-haacker replied:
I tried to use either `return type(self)(..., fastpath=True)` with attrs or the `self._contruct_direct()`, like it is used in `Dataset()`, but neither take attrs. From the regular contructor in the `fastpath=False` logic branch, I see that attrs are given to `variable` and not assigned directly.

I think alternatives are setting `fastpath=False` or using `.assign_attrs()`.

headtr1ck replied:
Ah, didn't see that. In this case let's keep the fast path to not get s performance regression.

headtr1ck reviewed (approved):
Will approve to not delay this further, but I think there is still a flaw in the logic that should be fixed.

j-haacker commented:
I fixed the issue @headtr1ck  raised, that replacing a variable and attributes simultaneously would "skip" the variable.

I added tests for `_replace`. However:

1. I noticed the methods let by one "_" don't have tests. Are they tested somewhere else or not meant to be tested?
2. I'm not sure about the desired replacement logic if a variable is replaced that has attributes.

(2) I made the assumption that if a variable and attributes are replaced simultaneously, the passed attributes override the variable attributes. And further, if the replacement variable contains attributes they should replace any original attributes. I made both explicit in the test. I'm happy to change the behavior or add a warning.

max-sixty commented:
Great — we do try and test private methods, but sometimes assume they're implicitly tested. When making a change to them, adding a test like you most recently did makes a lot of sense.

Thanks a lot @j-haacker ! Appreciate you making the underlying code better as well as fixing the immediate problem!
